pkg/endpoint/bpf.go

```diff
@@ -962,6 +1014,63 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
 	// returns changes that need to be applied to the Endpoint's bpf policy map.
 	closer, changes := e.desiredPolicy.ConsumeMapChanges()
 	defer closer()
+	var deleteLen int
+	for k := range changes.Deletes {
+		if _, ok := changes.Adds[k]; !ok {
+			deleteLen++
+		}
+	}
+	// If the desiredPolicy will be larger than the BPF maximum after
+	// the changes we need to reset the consumed changes and either
+	// lockdown or return an inevitable error (which would happen on
+	// insertion if we proceed).
+	if e.desiredPolicy.GetPolicyMap().Len()+len(changes.Adds)-deleteLen > policymap.MaxEntries {
+		// Revert the map changes that were just consumed.
+		e.desiredPolicy.RevertChanges(changes)
+		// We only need to go through the mechanics of
+		// lockdown once.
+		if !e.lockdown {
+			// Avoid an insertion error further down and return an
+			// error now.
+			if !option.Config.EnableEndpointLockdownOnPolicyOverflow {
+				// Do not return ErrPolicyEntryMaxExceeded as it
+				// implies that option.Config.EnableEndpointLockdownOnPolicyOverflow
+				// is "true".
+				return fmt.Errorf("policy map exceeds the map max entries limit")
+			}
+
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.EndpointID: e.ID,
+			}).Warnf("The policy map exceeds the max entries limit, %s is enabled, locking the endpoint down.",
+				option.EnableEndpointLockdownOnPolicyOverflow)
+
+			if err := e.endpointPolicyLockdown(); err != nil {
+				e.getLogger().WithFields(logrus.Fields{
+					logfields.EndpointID: e.ID,
+				}).WithError(err).Error("Failed to lockdown endpoint:" +
+					"The endpoint may be compromised. Consider quarantining or shutting down this node.")
+				return err
+			}
+			e.lockdown = true
+		}
+		e.updatePolicyMapPressureMetric(float64(e.desiredPolicy.GetPolicyMap().Len() + len(changes.Adds) - deleteLen))
+		return ErrPolicyEntryMaxExceeded
+	}
+
+	if e.lockdown {
```

Comment:
The `endpoint.policyMap` is replaced in lockdown, given a temporary pin path, and then pinned to the correct path for the datapath. Coming out of lockdown `syncPolicyMap` syncs `desiredPolicy` (which has all of the policies) and `realizedPolicy` (which, during lockdown, has 6 deny all policies in it). The logic inserts all of the `desiredPolicy` entries into `realizedPolicy` and the change state is inserted into the datapath map. No clean up is necessary. The reason for replacing the "old" map for lockdown is that it is faster to simply replace the map rather than do delete syscalls for all of the entries (especially for large maps). The inverse is not true; when coming out of lockdown the map must be rebuilt, so no replacement swap can work to make it faster, the entries must be reinserted. I would like to start utilizing some of the batch ops in this package (which would definitely help in this situation).

Reply:
Ok I think I follow: if we come out of lockdown we reuse the lockdown map and fill it with actual policies correct? But then the name of the map will be wrong right? It will have the `_lockdown` suffix since we are not recreating the policy map, and the lockdown mapped was named with the tmp pin path.

Reply:
No. During lockdown it is created `Unpinned` with a `_lockdown` suffix, but when it is pinned the file is changed (on line 1199) `e.policyMap.Pin(e.policyMapPath())` so it has the correct name during lockdown (otherwise it wouldn't work).

Reply:
Sorry for the long wait... I think I am talking about the name stored with the bpf ressource, and you are talking about the pin path. I understand the pin path handling is correct, however the map name will keep the `_lockdown` suffix if its not replaced. My suggestion would be that we just don't use the `_lockdown` suffix at all as it doesn't really have any use IIUC. Unfortunately, I also can't prove my point using evidence, `bpftool `never displays more than 15 characters of a map name. I will approve regardless to not block this PR any longer and I don't think it has any bad side effects, but I am pretty sure we are permanently changing the map name when locking down an endpoint, even if the pin path stays the same.